libraries/lib-basic-ui/BasicUI.h

```diff
@@ -0,0 +1,289 @@
+   { parent = parent_; return std::move(*this); }
+
+   MessageBoxOptions &&Caption(TranslatableString caption_) &&
+   { caption = std::move(caption_); return std::move(*this); }
+
+   MessageBoxOptions &&IconStyle(Icon style) &&
+   { iconStyle = style; return std::move(*this); }
+
+   MessageBoxOptions &&ButtonStyle(Button style) &&
+   { buttonStyle = style; return std::move(*this); }
+
+   //! Override the usual defaulting to Yes; affects only the YesNo case
+   MessageBoxOptions &&DefaultIsNo() &&
+   { yesOrOkDefaultButton = false; return std::move(*this); }
+
+   MessageBoxOptions &&CancelButton() &&
+   { cancelButton = true; return std::move(*this); }
+
+   //! Center the dialog on its parent window, if any
+   MessageBoxOptions &&Centered() &&
+   { centered = true; return std::move(*this); }
+
+   //! @}
+
+   WindowPlacement *parent{ nullptr };
+   TranslatableString caption{ DefaultCaption() };
+   Icon iconStyle{ Icon::None };
+   Button buttonStyle{ Button::Default };
+   bool yesOrOkDefaultButton{ true };
+   bool cancelButton{ false };
+   bool centered{ false };
+};
+
+enum class MessageBoxResult : int {
+   None, //!< May be returned if no Services are installed
+   Yes,
+   No,
+   Ok,
+   Cancel,
+};
+
+enum ProgressDialogOptions : unsigned {
+   ProgressShowStop            = (1 << 0),
+   ProgressShowCancel          = (1 << 1),
+   ProgressHideTime            = (1 << 2),
+   ProgressConfirmStopOrCancel = (1 << 3),
+};
+
+enum class ProgressResult : unsigned
+{
+   Cancelled = 0, //<! User says that whatever is happening is undesirable and shouldn't have happened at all
+   Success,       //<! User says nothing, everything works fine, continue doing whatever we're doing
+   Failed,        //<! Something has gone wrong, we should stop and cancel everything we did
+   Stopped        //<! Nothing is wrong, but user says we should stop now and leave things as they are now
+};
+
+//! Abstraction of a progress dialog with well defined time-to-completion estimate
+class BASIC_UI_API ProgressDialog
+{
+public:
+   virtual ~ProgressDialog();
+
+   // Update the bar and poll for clicks
+   virtual ProgressResult Poll(
+      unsigned long long numerator,
+      unsigned long long denominator,
+      const TranslatableString &message = {}) = 0;
+};
+
+//! Abstraction of a progress dialog with undefined time-to-completion estimate
+class BASIC_UI_API GenericProgressDialog
+{
+public:
+   virtual ~GenericProgressDialog();
+   //! Give some visual indication of progress
+   virtual void Pulse() = 0;
+};
+
+//! @}
+
+//! Abstract class defines a few user interface services, not mentioning particular toolkits
+/*! The intention is that the application supplies a concrete implementation at
+ startup.  Most code will not use this class directly, but call the inline
+ functions that follow. */
+class BASIC_UI_API Services {
+public:
+   virtual ~Services();
+   virtual void DoCallAfter(const Action &action) = 0;
```

Comment:
I think this can be placed into more generic Job/Task scheduler, this might be useful outside UI-related code

Reply:
Maybe so.  Maybe that argues for factoring a base class (still abstract) which Services extends.

It's not needed yet for the purposes of the present pull request.